### ISSUE TITLE
FIDO2 generated signature is not correct

### ISSUE THREAD

**Issue by eadwu** [Yubico/yubikit-android#152]
I'm having trouble authenticating the responses from the yubikey. I used https://webauthn.passwordless.id/demos/playground to verify

```
{
  "clientExtensionResults": {},
  "id": "Z1mCLKNx8OaAdpGC8Gzh4YK9487jEJh1o68mxboIqcbGPCjT5-_YvxmG-b5C34p8",
  "rawId": "Z1mCLKNx8OaAdpGC8Gzh4YK9487jEJh1o68mxboIqcbGPCjT5-_YvxmG-b5C34p8",
  "type": "public-key",
  "authenticatorAttachment": "cross-platform",
  "response": {
    "authenticatorData": "T7IIVvJKaufa_CeBCQrIR3rm4r0HJmAjbMYUxvt8LqAFAAAAAw==",
    "clientDataJSON": "eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoiZVVSQUR2QURMVmtxRlY4X0RlMTJYUFdtIiwib3JpZ2luIjoiaHR0cHM6Ly93ZWJhdXRobi5wYXNzd29yZGxlc3MuaWQiLCJjcm9zc09yaWdpbiI6ZmFsc2V9",
    "signature": "MEUCIQDIBZlIn9s2laUDX7R45b1TA8FFISUe3T1SyUPAtrSV_QIgI4-otKgR8v6drJFtRuhluvaAzwuY8NH8tCWA2q_kh6c=",
    "userHandle": "ZDQ4NjA0MGYtNDllMi00NTFhLWE5ODktOWY0ZTRiOWIyNTc2"
  }
}
```

Everything else matches from the registration, `userHandle`, and the JSON of `clientDataJSON`. I guess I'm not too sure about `authenticatorData`

**Comment by AdamVe** [comment]
Hi @eadwu, could you please share the code which you use to perform the FIDO2 authentication.

**Comment by eadwu** [comment]
```kotlin
            val clientDataJSON = JSONObject()
            clientDataJSON.put("type", "webauthn.get")
            clientDataJSON.put("challenge", b64Encode(req.challenge))
            clientDataJSON.put("origin", origin)
            clientDataJSON.put("crossOrigin", false)

            val opts = com.yubico.yubikit.fido.webauthn.PublicKeyCredentialRequestOptions(
                req.challenge,
                req.timeout,
                req.rpId,
                listOf(publicKeyDescriptor),
                UserVerificationRequirement.PREFERRED,
                null,
            )

            val publicKeyCredential = client.getAssertion(
                clientDataJSON.toString().toByteArray(), opts, req.rpId,
                "1234".toCharArray(), null
            )
```

Logs from Yubikey
```
com.yubico....ClientPin D  Getting shared secret
com.yubico...ap2Session D  clientPin for pinUvAuthProtocol=2,subCommand=2,keyAgreement=null,pinUvAuthParam=null,newPinEnc=null,pinHashEnc=null,permissions=null,rpId=null
com.yubico...Connection V  sent: 80 10 00 00 06 06 a2 01 02 02 02 
com.yubico...Connection V  received: 00 a1 01 a5 01 02 03 38 18 20 01 21 58 20 5d dc f6 b3 3b b9 4a fe 17 de ab 44 25 26 7e 77 e7 60 97 94 d3 1e 63 c9 4a 6f a9 f4 35 5f a3 dc 22 58 20 cf da bd a6 45 ff 27 42 26 d2 62 ec 94 9e 5e 0e 8e 67 ac 22 bf 68 61 6a f9 94 62 40 5d 81 53 63 90 00 
com.yubico....ClientPin D  Getting PIN token
com.yubico...ap2Session D  clientPin for pinUvAuthProtocol=2,subCommand=9,keyAgreement={-1=1, 1=2, -2=[B@49c95b7, -3=[B@ed3b124, 3=-25},pinUvAuthParam=null,newPinEnc=null,pinHashEnc=[-16, 13, -42, -8, 99, 63, -97, -95, -13, -37, 62, 82, 87, -110, -78, 100, -65, 26, 87, -38, -90, -39, -41, 96, 59, 30, -84, 84, -86, 72, 11, 114],permissions=2,rpId=webauthn.passwordless.id
com.yubico...Connection V  sent: 80 10 00 00 95 06 a6 01 02 02 09 03 a5 01 02 03 38 18 20 01 21 58 20 f3 5b 43 29 53 b9 a6 2b d6 1c 67 d0 32 f4 13 bb 48 a1 fb cb 2c 8f 8e 7e fc 32 c4 80 4d 71 45 1f 22 58 20 fe 3b a8 9d b5 c4 d3 7d 92 b5 98 96 97 0c 63 70 f2 75 bb 81 e3 1c d4 55 a9 d5 c1 bd 78 ff d4 ba 06 58 20 f0 0d d6 f8 63 3f 9f a1 f3 db 3e 52 57 92 b2 64 bf 1a 57 da a6 d9 d7 60 3b 1e ac 54 aa 48 0b 72 09 02 0a 78 18 77 65 62 61 75 74 68 6e 2e 70 61 73 73 77 6f 72 64 6c 65 73 73 2e 69 64 
com.yubico...Connection V  received: 00 a1 02 58 30 43 5b 40 3e ca 3d 69 52 6c e9 99 5f 50 4e 63 f5 8d a2 fa 58 15 29 c1 03 83 72 b2 a5 d0 85 6f f2 57 85 c1 65 18 cd 0a fc 74 56 b6 45 cb 5f 98 b9 90 00 
com.yubico....ClientPin D  Got PIN token for permissions: 2, permissions rpID: webauthn.passwordless.id
com.yubico...ap2Session D  getAssertions for rpId=webauthn.passwordless.id,clientDataHash=[-115, 118, -37, -121, -5, 127, -3, -39, 90, 21, 7, -45, 20, 56, -123, 77, -24, -61, -98, -22, 58, -99, -71, 74, 70, -22, 71, 125, 122, -58, 94, 84],allowList=[{id=[B@5f3948d, type=public-key}],extensions=null,options=null,pinUvAuthParam=[-86, -46, 45, -46, 72, -77, 96, 71, -114, -74, -25, -75, -117, -39, -28, 58, -36, -39, 34, 42, 17, 107, -115, 52, 115, -89, -27, -8, 27, -55, -116, -18],pinUvAuthProtocol=2,state={}
com.yubico...Connection V  sent: 80 10 00 00 ad 02 a5 01 78 18 77 65 62 61 75 74 68 6e 2e 70 61 73 73 77 6f 72 64 6c 65 73 73 2e 69 64 02 58 20 8d 76 db 87 fb 7f fd d9 5a 15 07 d3 14 38 85 4d e8 c3 9e ea 3a 9d b9 4a 46 ea 47 7d 7a c6 5e 54 03 81 a2 62 69 64 58 30 62 2f 71 fb d6 81 8b 49 04 70 77 e1 ee 51 04 9b 43 f8 a4 98 aa 81 ed 1a 6c ce ca 92 0c c4 1d cc d9 b1 37 fe 33 bc ea da c7 6b 00 af ae b7 99 6f 64 74 79 70 65 6a 70 75 62 6c 69 63 2d 6b 65 79 06 58 20 aa d2 2d d2 48 b3 60 47 8e b6 e7 b5 8b d9 e4 3a dc d9 22 2a 11 6b 8d 34 73 a7 e5 f8 1b c9 8c ee 07 02 
com.yubico...Connection V  received: 00 a4 01 a2 62 69 64 58 30 62 2f 71 fb d6 81 8b 49 04 70 77 e1 ee 51 04 9b 43 f8 a4 98 aa 81 ed 1a 6c ce ca 92 0c c4 1d cc d9 b1 37 fe 33 bc ea da c7 6b 00 af ae b7 99 6f 64 74 79 70 65 6a 70 75 62 6c 69 63 2d 6b 65 79 02 58 25 4f b2 08 56 f2 4a 6a e7 da fc 27 81 09 0a c8 47 7a e6 e2 bd 07 26 60 23 6c c6 14 c6 fb 7c 2e a0 05 00 00 00 1a 03 58 46 30 44 02 20 72 3e 2f cc 3d 80 6d 33 71 38 8a b7 15 a6 f0 0a ec b5 72 05 8b b7 34 e9 00 d6 55 df 06 db 40 26 02 20 49 72 7b 1c 2e 1b 88 e2 02 ea 55 c6 b5 5c c0 a5 b9 7d a6 c4 98 17 90 08 54 1a b3 16 c8 8d 5a 98 04 a1 62 69 64 58 24 65 34 31 39 35 36 65 38 2d 66 31 66 63 2d 34 35 63 63 2d 38 65 36 66 2d 32 36 63 36 38 61 35 38 34 35 66 61 90 00 
com.yubico...ap2Session I  Authenticator returned 1 assertions.
fido2                   D  got response {"authenticatorAttachment":"cross-platform","response":{"userHandle":"ZTQxOTU2ZTgtZjFmYy00NWNjLThlNmYtMjZjNjhhNTg0NWZh","clientDataJSON":"eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoidlEzOVBYZXFVQkpVeVFaQWp1OXB1T0pncUNJWXg2ZTAxWm1Qb0UxWmJqcyIsIm9yaWdpbiI6Imh0dHBzOlwvXC93ZWJhdXRobi5wYXNzd29yZGxlc3MuaWQiLCJjcm9zc09yaWdpbiI6ZmFsc2V9","signature":"MEQCIHI-L8w9gG0zcTiKtxWm8ArstXIFi7c06QDWVd8G20AmAiBJcnscLhuI4gLqVca1XMCluX2mxJgXkAhUGrMWyI1amA","authenticatorData":"T7IIVvJKaufa_CeBCQrIR3rm4r0HJmAjbMYUxvt8LqAFAAAAGg"},"rawId":"Yi9x-9aBi0kEcHfh7lEEm0P4pJiqge0abM7KkgzEHczZsTf-M7zq2sdrAK-ut5lv","id":"Yi9x-9aBi0kEcHfh7lEEm0P4pJiqge0abM7KkgzEHczZsTf-M7zq2sdrAK-ut5lv","type":"public-key"}
```

**Comment by eadwu** [comment]
Still broken on 2.7.0, though it didn't look like it changed anything regarding fido2 to begin with.

**Comment by AdamVe** [comment]
Hi, sorry for late reply. I checked out your code but cannot run it. I tried also with my test apps where I successfully executed makeCredential and getAssertions with yubikit-android (on Pixel 6, Android 14, USB and NFC) on https://webauthn.passwordless.id/demos/playground and as well was able to authenticate with a credential created through a desktop browser.

Could you explain why you think that the assertion signature is wrong?

[This figure  ](https://www.w3.org/TR/webauthn-2/#fig-signature) shows that the signature is created from concatenated `authenticatorData` and `clientDataHash` using the credential’s private key. `authenticatorData` will be different for every run because of the internal values of the authenticator, like signature counter - paste your authenticator data [here](https://demo.yubico.com/webauthn-developers) to see what it contains.

**Comment by eadwu** [comment]
I figured it out through the yubikey playground ... since I'm getting my ground of truth for the WebAuthn query from Android's CredentialManager before proxying that to the Yubikey, I was trusting it to be representative of the original query. 

It looks like while processing, the challenge is replaced to be something else before being replaced in the response as well, so the challenge I was using was wrong. 

Android works around this by providing the precomputed `clientDataHash` which was computed using the correct challenge. Stubbing out a function in a locally compared jar that injects that `clientDataHash` makes the output signature correct.

tldr Android hijacked request and I never see the original `challenge` value in the app but the output is correct. Need to provide the given `clientDataHash` and `challenge` will be fixed by Android.

**Comment by AdamVe** [comment]
Thanks for the details, If you need any help with the sdk don't hesitate to let us know 👍🏻

**Comment by Ajeebiw** [comment]
> شكرًا على التفاصيل، إذا كنت بحاجة إلى أي مساعدة مع مجموعة أدوات التطوير البرمجية فلا تتردد في إخبارنا بذلك 👍🏻

![52175734671_912c2172fc_vm_exported_0](https://github.com/user-attachments/assets/b51884ad-82b0-4816-9c81-541afdb52bfc)
